include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,547 @@
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_PARALLEL_BACKEND_SYCL_HISTOGRAM_H
+#define _ONEDPL_PARALLEL_BACKEND_SYCL_HISTOGRAM_H
+
+#include <algorithm>
+#include <cstdint>
+#include <type_traits>
+
+#include "sycl_defs.h"
+#include "parallel_backend_sycl_utils.h"
+#include "execution_sycl_defs.h"
+#include "unseq_backend_sycl.h"
+#include "utils_ranges_sycl.h"
+#include "oneapi/dpl/internal/histogram_binhash_utils.h"
+#include "oneapi/dpl/internal/async_impl/async_impl_hetero.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __par_backend_hetero
+{
+
+//no boost baseline
+template <typename _BinHash>
+struct __binhash_SLM_wrapper
+{
+    _BinHash __bin_hash;
+    __binhash_SLM_wrapper(_BinHash __bin_hash) : __bin_hash(__bin_hash) {}
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value) const
+    {
+        return __bin_hash.get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value) const
+    {
+        return __bin_hash.is_valid(::std::forward<_T2>(__value));
+    }
+
+    inline ::std::size_t
+    get_required_SLM_memory() const
+    {
+        return 0;
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
+    {
+    }
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value, void* __boost_mem) const
+    {
+        return get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value, void* __boost_mem) const
+    {
+        return is_valid(::std::forward<_T2>(__value));
+    }
+};
+
+template <typename _Range>
+struct __binhash_SLM_wrapper<oneapi::dpl::__internal::__custom_range_binhash<_Range>>
+{
+    using _BinHashType = typename oneapi::dpl::__internal::__custom_range_binhash<_Range>;
+    _BinHashType __bin_hash;
+
+    using __boundary_type = typename _BinHashType::__boundary_type;
+
+    __binhash_SLM_wrapper(_BinHashType __bin_hash) : __bin_hash(__bin_hash) {}
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value) const
+    {
+        return __bin_hash.get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value) const
+    {
+        return __bin_hash.is_valid(::std::forward<_T2>(__value));
+    }
+
+    inline ::std::size_t
+    get_required_SLM_memory()
+    {
+        return sizeof(__boundary_type) * __bin_hash.__boundaries.size();
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
```

Comment:
`void*` and further `static_cast` technic doesn't look a C++ type safety approach... Probably, can we avoid that?

Reply:
My intention here is to allow `__binhash_SLM_wrapper<T>` to be called generically by the kenels without needing to know how the extra SLM memory is going to be used by the individual bin hash structure.  For evenly divided bins, it is unused (and size=0).  For custom ranges, it is used to cache the boundaries.  In theory, one could use the SLM memory for something else if they add another histogram API in the future.  

Is there a better way than void* to handle this memory but keep it a generic call?

Reply:
I've refactored this a bit, letting binhash dictate the type of the extra memory.  I've also split out the extra memory to a separate block from the local histogram bin storage.  This allows for better code and not having to cast pointers around.

There may be a minor performance hit by splitting the SLM allocations this but it is better code.  I'll investigate to make sure we don't lose too much.